{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Constants for the unicode characters used in sentence detection\n",
    "PUNCT_INDICATOR = \"╚\"\n",
    "ELLIPSIS_INDICATOR = \"╦\"\n",
    "ABBREVIATOR = \"╞\"\n",
    "NUM_INDICATOR = \"╟\"\n",
    "MULT_PERIOD = \"╔\"\n",
    "SPECIAL_PERIOD = \"╩\"\n",
    "QUESTION_IN_QUOTE = \"╤\"\n",
    "EXCLAMATION_INDICATOR = \"╥\"\n",
    "BREAK_INDICATOR = \"\\uF050\"\n",
    "DOT = \"\\uF051\"\n",
    "COMMA = \"\\uF052\"\n",
    "SEMICOLON = \"\\uF053\"\n",
    "QUESTION = \"\\uF054\"\n",
    "EXCLAMATION = \"\\uF055\"\n",
    "PROTECTION_MARKER_OPEN = \"\\uF056\"\n",
    "PROTECTION_MARKER_CLOSE = \"\\uF057\"\n",
    "PROTECT_CHAR = \"ↈ\"\n",
    "BREAK_CHAR = \"ↇ\"\n",
    "\n",
    "reverse_dict = {\n",
    "    DOT: \".\",\n",
    "    SEMICOLON: \";\",\n",
    "    QUESTION: \"?\",\n",
    "    EXCLAMATION: \"!\",\n",
    "    PROTECTION_MARKER_OPEN: \"\",\n",
    "    PROTECTION_MARKER_CLOSE: \"\",\n",
    "    ABBREVIATOR: \".\",\n",
    "    NUM_INDICATOR: \".\",\n",
    "    MULT_PERIOD: \".\",\n",
    "    QUESTION_IN_QUOTE: \"?\",\n",
    "    EXCLAMATION_INDICATOR: \"!\",\n",
    "    ELLIPSIS_INDICATOR: \"...\",\n",
    "}\n",
    "\n",
    "\n",
    "class Formater:\n",
    "    \"\"\"A class used for sentence detection.\"\"\"\n",
    "\n",
    "    _PREPOSITIVE_ABBREVIATIONS = [\"dr\", \"mr\", \"ms\", \"mt\", \"st\"]\n",
    "    _NUMBER_ABBREVIATIONS = [\"no\", \"px\"]\n",
    "\n",
    "    _pabb = \"(?:\" + \"|\".join(_PREPOSITIVE_ABBREVIATIONS) + \")\"\n",
    "    _nubb = \"(?:\" + \"|\".join(_NUMBER_ABBREVIATIONS) + \")\"\n",
    "\n",
    "    _number_rules = [\n",
    "        \"(?<=\\d)\\.(?=\\d)\",\n",
    "        \"\\.(?=\\d)\",\n",
    "        \"(?<=\\d)\\.(?=\\S)\",\n",
    "        \"(?<=^\\d)\\.(?=(\\s\\S)|\\))\",\n",
    "        \"(?<=^\\d\\d)\\.(?=(\\s\\S)|\\))\",\n",
    "    ]\n",
    "\n",
    "    _ditc_abbr_rules = [\n",
    "        rf\"(?i)(?<=\\s{_pabb})\\.(?=\\s)|(?<=^{_pabb})\\.(?=\\d+)\",\n",
    "        rf\"(?i)(?<=\\s{_pabb})\\.(?=\\d+)|(?<=^{_pabb})\\.(?=\\d+)\",\n",
    "        rf\"(?i)(?<=\\s{_nubb})\\.(?=s\\\\d)|(?<=^{_nubb})\\.(?=s\\\\d)\",\n",
    "        rf\"(?i)(?<=\\s{_nubb})\\.(?=:\\s+\\()|(?<=^{_nubb})\\.(?=\\s+\\()\",\n",
    "    ]\n",
    "\n",
    "    _std_abbre_rules = [\n",
    "        \"\\.(?='s\\s)|\\.(?='s\\$)|\\.(?='s\\Z)\",\n",
    "        \"(?<=Co)\\.(?=\\sKG)\",\n",
    "        \"(?<=^[A-Z])\\.(?=\\s)\",\n",
    "        \"(?<=\\s[A-Z])\\.(?=\\s)\",\n",
    "    ]\n",
    "\n",
    "    _specials_abbr_rules = [\n",
    "        r\"\\b[a-zA-Z](?:\\.[a-zA-Z])+(?:\\.(?!\\s[A-Z]))*\",\n",
    "        r\"(?i)p\\.m\\.*\",\n",
    "        r\"(?i)a\\.m\\.\",\n",
    "    ]\n",
    "\n",
    "    _punctuations_rules = [\"(?<=\\S)[!\\?]+(?=\\s|\\Z|\\$)\"]\n",
    "\n",
    "    _multiple_periods_rules = [\"(?<=\\w)\\.(?=\\w)\"]\n",
    "\n",
    "    _geolocations_rules = [\"(?<=[a-zA-z]°)\\.(?=\\s*\\d+)\"]\n",
    "\n",
    "    _ellipsis_rules = [\n",
    "        \"\\.\\.\\.(?=\\s+[A-Z])\",\n",
    "        \"(?<=\\S)\\.{3}(?=\\.\\s[A-Z])\",\n",
    "    ]\n",
    "\n",
    "    _between_punctuations_rules = [\n",
    "        \"(?<=\\s|^)'[\\w\\s?!\\.,|'\\w]+'(?:\\W)\",\n",
    "        '\"[\\w\\s?!\\.,]+\"',\n",
    "        \"\\[[\\w\\s?!\\.,]+\\]\",\n",
    "        \"\\([\\w\\s?!\\.,]+\\)\",\n",
    "    ]\n",
    "\n",
    "    _queation_mark_inquation_rules = [\"\\?(?=('|\\\"))\"]\n",
    "\n",
    "    _exclamations_rules = [\"\\!(?=('|\\\"))\", \"\\!(?=\\,\\s[a-z])\", \"\\!(?=\\s[a-z])\"]\n",
    "\n",
    "    _basic_rules = {DOT: \"\\.\", SEMICOLON: \";\"}\n",
    "\n",
    "    def _replace_rules(self, content: str, simbol: str, rules: list):\n",
    "        for rule in rules:\n",
    "            content = re.sub(rule, simbol, content)\n",
    "        return content\n",
    "\n",
    "    def _replace_break_and_symbolic(self, content: str, rules: dict):\n",
    "        for key, rule in rules.items():\n",
    "            content = re.sub(rule, key + BREAK_INDICATOR, content)\n",
    "        return content\n",
    "\n",
    "    def _replace_with_all_simbols(self, content: str, simbol: str, rules: list):\n",
    "        return self._replace_rules(content, simbol, rules)\n",
    "\n",
    "    def _replace_with_all_simbols_and_breaks(\n",
    "        self, content: str, simbol: str, rules: list\n",
    "    ):\n",
    "        return self._replace_rules(content, simbol + BREAK_INDICATOR, rules)\n",
    "\n",
    "    def _replace_with_protect_breaks(self, content: str, rules: list):\n",
    "        return self._replace_rules(\n",
    "            content, PROTECTION_MARKER_OPEN + \"\\g<0>\" + PROTECTION_MARKER_CLOSE, rules\n",
    "        )\n",
    "\n",
    "    def _replace_with_appendl(self, content: str, simbol: str, rules: list):\n",
    "        return self._replace_rules(content, \"\\g<0>\" + simbol, rules)\n",
    "\n",
    "    def formatNumbers(self, text: str):\n",
    "        return self._replace_with_all_simbols(\n",
    "            text, NUM_INDICATOR, Formater._number_rules\n",
    "        )\n",
    "\n",
    "    def formatAbbreviations(self, text: str):\n",
    "        # There are logic in open source that could be wrong\n",
    "        replacement = self._replace_with_protect_breaks(\n",
    "            text, Formater._specials_abbr_rules\n",
    "        )\n",
    "        replacement = self._replace_with_all_simbols(\n",
    "            replacement, ABBREVIATOR, Formater._std_abbre_rules\n",
    "        )\n",
    "        return self._replace_with_all_simbols(\n",
    "            replacement, ABBREVIATOR, Formater._ditc_abbr_rules\n",
    "        )\n",
    "\n",
    "    def formatPunctuations(self, text: str):\n",
    "        return self._replace_with_appendl(\n",
    "            text, BREAK_INDICATOR, Formater._punctuations_rules\n",
    "        )\n",
    "\n",
    "    def formatEllipsisRules(self, text):\n",
    "        return self._replace_with_all_simbols_and_breaks(\n",
    "            text, ELLIPSIS_INDICATOR, Formater._punctuations_rules\n",
    "        )\n",
    "\n",
    "    def formatMultiplePeriods(self, text: str):\n",
    "        return self._replace_with_all_simbols(\n",
    "            text, MULT_PERIOD, Formater._multiple_periods_rules\n",
    "        )\n",
    "\n",
    "    def formatGeoLocations(self, text: str):\n",
    "        return self._replace_with_all_simbols(\n",
    "            text, MULT_PERIOD, Formater._geolocations_rules\n",
    "        )\n",
    "\n",
    "    def formatBetweenPunctuations(self, text: str):\n",
    "        # There are logic in open source that could be wrong\n",
    "        return text\n",
    "\n",
    "    def formatQuotationMarkInQuotation(self, text: str):\n",
    "        return self._replace_with_all_simbols(\n",
    "            text, QUESTION_IN_QUOTE, Formater._queation_mark_inquation_rules\n",
    "        )\n",
    "\n",
    "    def formatExclamationPoint(self, text: str):\n",
    "        return self._replace_with_all_simbols(\n",
    "            text, EXCLAMATION_INDICATOR, Formater._exclamations_rules\n",
    "        )\n",
    "\n",
    "    def formatBasicBreakers(self, text: str):\n",
    "        return self._replace_break_and_symbolic(text, Formater._basic_rules)\n",
    "\n",
    "    def generate_protecedContent(self, content):\n",
    "        content_modified = self.formatNumbers(content)\n",
    "        content_modified = self.formatAbbreviations(content_modified)\n",
    "        content_modified = self.formatPunctuations(content_modified)\n",
    "        content_modified = self.formatGeoLocations(content_modified)\n",
    "        content_modified = self.formatMultiplePeriods(content_modified)\n",
    "        content_modified = self.formatEllipsisRules(content_modified)\n",
    "        content_modified = self.formatBetweenPunctuations(content_modified)\n",
    "        content_modified = self.formatQuotationMarkInQuotation(content_modified)\n",
    "        content_modified = self.formatExclamationPoint(content_modified)\n",
    "        content_modified = self.formatBasicBreakers(content_modified)\n",
    "        return content_modified"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"This module contains the PubmedPMID class for generating a JSON object for a PubMed article.\n",
    "The JSON object contains the abstract of the article and the meta information of the article such as the article\n",
    "title, authors, journal info, etc. The JSON object also contains the article splits\n",
    "which are the abstract split into paragraphs of max_split_token_length &\n",
    "min_split_token_length tokens with an overlap of sentence_overlap sentences.\n",
    "\n",
    "Example:\n",
    "    >>> from pubmed_pmid import PubmedPMID\n",
    "    >>> pubmed_pmid = PubmedPMID()\n",
    "    >>> json_output = pubmed_pmid.build_pubmed_json(xml_path=\"data/36464825.xml\")\n",
    "    >>> print(json_output)\n",
    "\n",
    "    or with large gz files:\n",
    "    >>> from pubmed_pmid import PubmedPMID\n",
    "    >>> pubmed_pmid = PubmedPMID()\n",
    "    >>> dicts_out = pubmed.parse_pubmed_xml_iter(\"data/pubmed23n1166.xml.gz\")\n",
    "\n",
    "    >>> for article in dicts_out:\n",
    "    >>>     if article is None:\n",
    "    >>>         continue\n",
    "    >>>     pmid = article[\"pmid\"]\n",
    "    >>>     with open(f\"./data/pubmed/pubmed_{pmid}.json\", \"w\") as f:\n",
    "    >>>         json.dump(article, f)\n",
    "\n",
    "\n",
    "Some codes are adapted from https://github.com/titipata/pubmed_parser\n",
    "\"\"\"\n",
    "import re\n",
    "from urllib.parse import quote\n",
    "from itertools import chain\n",
    "\n",
    "import gzip\n",
    "\n",
    "import requests\n",
    "from lxml import etree\n",
    "\n",
    "from bs4 import BeautifulSoup\n",
    "import tiktoken\n",
    "\n",
    "\n",
    "class PubmedPMID:\n",
    "    \"\"\"Class for generating a JSON object for a PubMed article. The JSON object contains\n",
    "    the abstract of the article and the meta information of the article such as the article\n",
    "    title, authors, journal info, etc. The JSON object also contains the article splits\n",
    "    which are the abstract split into paragraphs of max_split_token_length & \n",
    "    min_split_token_length tokens with an overlap of sentence_overlap sentences.\n",
    "\n",
    "    Args:\n",
    "        max_split_token_length (int, optional): The maximum number of tokens in a split. Defaults to 500.\n",
    "        min_split_token_length (int, optional): The minimum number of tokens in a split. Defaults to 100.\n",
    "        sentence_overlap (int, optional): The number of sentences to overlap between splits. Defaults to 2.\n",
    "        get_citation_count (bool, optional): Whether to get the citation count of the article. Defaults to False.\n",
    "        get_journal_ranking (bool, optional): Whether to get the ranking of the journal. Defaults to False.\n",
    "\n",
    "\n",
    "    Attributes:\n",
    "        json_output (dict): The JSON object for the article\n",
    "        max_split_token_length (int): The maximum number of tokens in a split\n",
    "        min_split_token_length (int): The minimum number of tokens in a split\n",
    "        sentence_overlap (int): The number of sentences to overlap between splits\n",
    "        journal_ranking_dict (dict): A cache  dictionary of journal ranking info\n",
    "        get_citation_count (bool): Whether to get the citation count of the article\n",
    "        get_journal_ranking (bool): Whether to get the ranking of the journal\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(\n",
    "        self,\n",
    "        max_split_token_length: int = 500,\n",
    "        min_split_token_length: int = 100,\n",
    "        sentence_overlap: int = 2,\n",
    "        get_citation_count: bool = False,\n",
    "        get_journal_ranking: bool = False,\n",
    "    ):\n",
    "        self.max_split_token_length = max_split_token_length\n",
    "        self.sentence_overlap = sentence_overlap\n",
    "        self.min_split_token_length = min_split_token_length\n",
    "        self.citation_count_bool = get_citation_count\n",
    "        self.journal_ranking_bool = get_journal_ranking\n",
    "        self.journal_ranking_dict = {}\n",
    "\n",
    "    def get_pubmed_article_xml(self, pmid=None):\n",
    "        \"\"\"Get the PubMed article XML from the PMID and return it as a string\n",
    "        Args:\n",
    "            pmid (str): The PMID of the article\n",
    "        Returns:\n",
    "            str: The XML of the article\"\"\"\n",
    "\n",
    "        # check if the pmcid is valid\n",
    "        if pmid in [\"\", None]:\n",
    "            raise ValueError(\"PMCID is empty\")\n",
    "\n",
    "        params = {\n",
    "            \"db\": \"pubmed\",\n",
    "            \"id\": f\"[{pmid}]\",\n",
    "            \"rettype\": \"xml\",\n",
    "        }\n",
    "\n",
    "        try:\n",
    "            # Send the request to the NCBI\n",
    "            response = requests.get(\n",
    "                \"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi\",\n",
    "                params=params,\n",
    "                timeout=15,\n",
    "            )\n",
    "            response.raise_for_status()\n",
    "            xml_data = response.text\n",
    "            return xml_data\n",
    "\n",
    "        except requests.exceptions.RequestException as exception:\n",
    "            print(f\"An error occurred: {exception}\")\n",
    "            return None\n",
    "\n",
    "    def __stringify_children(self, node):\n",
    "        \"\"\"\n",
    "        Filters and removes possible Nones in texts and tails\n",
    "        ref: http://stackoverflow.com/questions/4624062/get-all-text-inside-a-tag-in-lxml\n",
    "        \"\"\"\n",
    "        parts = (\n",
    "            [node.text]\n",
    "            + list(chain(*([c.text, c.tail] for c in node.getchildren())))\n",
    "            + [node.tail]\n",
    "        )\n",
    "        return \"\".join(filter(None, parts)).strip()\n",
    "\n",
    "    def get_abstract(self, xml_tree):\n",
    "        \"\"\"Get the abstract text from the XML tree of the article\n",
    "        Args:\n",
    "            xml_tree (Element): The lxml node pointing to a medline document\n",
    "        Returns:\n",
    "            dict: A dictionary containing the abstract text, section title and section type\n",
    "        \"\"\"\n",
    "\n",
    "        article = xml_tree.find(\".//MedlineCitation/Article\")\n",
    "        category = \"Label\"\n",
    "        if article.find(\"Abstract/AbstractText\") is not None:\n",
    "            # parsing structured abstract\n",
    "            if len(article.findall(\"Abstract/AbstractText\")) > 1:\n",
    "                abstract_list = list()\n",
    "                for abstract in article.findall(\"Abstract/AbstractText\"):\n",
    "                    section = abstract.attrib.get(category, \"\")\n",
    "                    if section != \"UNASSIGNED\":\n",
    "                        abstract_list.append(\"\\n\")\n",
    "                        abstract_list.append(abstract.attrib.get(category, \"\") + \":\")\n",
    "\n",
    "                    section_text = self.__stringify_children(abstract)\n",
    "                    abstract_list.append(section_text)\n",
    "                abstract = \"\\n\".join(abstract_list)\n",
    "            else:\n",
    "                abstract = (\n",
    "                    self.__stringify_children(article.find(\"Abstract/AbstractText\"))\n",
    "                    or \"\"\n",
    "                )\n",
    "        elif article.find(\"Abstract\") is not None:\n",
    "            abstract = self.__stringify_children(article.find(\"Abstract\")) or \"\"\n",
    "        else:\n",
    "            abstract = \"\"\n",
    "\n",
    "        if abstract.strip() == \"\":\n",
    "            return None\n",
    "        else:\n",
    "            return {\n",
    "                \"text\": abstract,\n",
    "                \"section_title\": \"Abstract\",\n",
    "                \"section_type\": \"ABSTRACT\",\n",
    "            }\n",
    "\n",
    "    def get_article_ids(self, xml_tree):\n",
    "        \"\"\"Get the article IDs from the XML tree\n",
    "        Args:\n",
    "            xml_tree (Element): The lxml node pointing to a medline document\n",
    "        Returns:\n",
    "            list: A list of dictionaries containing the article IDs and their id types\n",
    "        \"\"\"\n",
    "\n",
    "        article_ids = []\n",
    "        article_id_list = xml_tree.findall(\".//PubmedData/ArticleIdList/ArticleId\")\n",
    "        if article_id_list is not None:\n",
    "            for article_id in article_id_list:\n",
    "                article_ids.append(\n",
    "                    {\n",
    "                        \"idtype\": article_id.attrib.get(\"IdType\", \"\"),\n",
    "                        \"value\": article_id.text,\n",
    "                    }\n",
    "                )\n",
    "        return article_ids\n",
    "\n",
    "    def get_keywords(self, xml_tree):\n",
    "        \"\"\"Get the keywords from the XML tree\n",
    "        Args:\n",
    "            xml_tree (Element): The lxml node pointing to a medline document\n",
    "        Returns:\n",
    "            str: A string of comma ``,`` separated keywords\n",
    "        \"\"\"\n",
    "        keywords = []\n",
    "        keyword_list = xml_tree.findall(\".//KeywordList/Keyword\")\n",
    "        if keyword_list is not None:\n",
    "            for keyword in keyword_list:\n",
    "                keywords.append(keyword.text or \"\")\n",
    "\n",
    "        return \", \".join(keywords)\n",
    "\n",
    "    def get_dates_history(self, xml_tree):\n",
    "        \"\"\"Get the dates from the XML tree\n",
    "        Args:\n",
    "            xml_tree (Element): The lxml node pointing to a medline document\n",
    "        Returns:\n",
    "            list: A list of dictionaries containing the date type, year, month and day\n",
    "        \"\"\"\n",
    "        dates = []\n",
    "        date_list = xml_tree.findall(\".//PubmedData/History/PubMedPubDate\")\n",
    "        if date_list is not None:\n",
    "            for date in date_list:\n",
    "                dates.append(\n",
    "                    {\n",
    "                        \"date_type\": date.attrib.get(\"PubStatus\", \"\"),\n",
    "                        \"year\": date.find(\"Year\").text,\n",
    "                        \"month\": date.find(\"Month\").text,\n",
    "                        \"day\": date.find(\"Day\").text,\n",
    "                    }\n",
    "                )\n",
    "        return dates\n",
    "\n",
    "    def get_journal_info(self, xml_tree):\n",
    "        \"\"\"Get the journal info from the XML tree\n",
    "        Args:\n",
    "            xml_tree (Element): The lxml node pointing to a medline document\n",
    "        Returns:\n",
    "            dict: A dictionary containing the journal name, \n",
    "                  journal abbreviation, ISSN and publication date\n",
    "        \"\"\"\n",
    "        journal_info = {}\n",
    "        journal = xml_tree.find(\".//MedlineCitation/Article/Journal\")\n",
    "\n",
    "        if journal is not None:\n",
    "            # Get the journal full name\n",
    "            title = journal.find(\"Title\")\n",
    "            if title is not None:\n",
    "                journal_info[\"fulljournalname\"] = title.text or \"\"\n",
    "\n",
    "            # Get the journal abbreviation name\n",
    "            abbrev = journal.find(\"ISOAbbreviation\")\n",
    "            if abbrev is not None:\n",
    "                journal_info[\"journal_abbrev\"] = abbrev.text or \"\"\n",
    "\n",
    "            # Get the journal ISSN\n",
    "            issn = journal.find(\"ISSN\")\n",
    "            if issn is not None:\n",
    "                journal_info[\"issn\"] = issn.text or \"\"\n",
    "\n",
    "            # Get the journal publication date\n",
    "            date = journal.find(\"JournalIssue/PubDate\")\n",
    "            if date is not None:\n",
    "                year = date.find(\"Year\")\n",
    "                year_text = \"\"\n",
    "                if year is not None:\n",
    "                    year_text = year.text\n",
    "\n",
    "                month = date.find(\"Month\")\n",
    "                month_text = \"\"\n",
    "                if month is not None:\n",
    "                    month_text = month.text\n",
    "\n",
    "                day = date.find(\"Day\")\n",
    "                day_text = \"\"\n",
    "                if day is not None:\n",
    "                    day_text = day.text\n",
    "                journal_info[\"pubdate\"] = f\"{year_text} {month_text} {day_text}\".strip()\n",
    "\n",
    "        return journal_info\n",
    "\n",
    "    def get_article_info(self, xml_tree):\n",
    "        \"\"\"Get the article info from the XML tree\n",
    "        Args:\n",
    "            xml_tree (Element): The lxml node pointing to a medline document\n",
    "        Returns:\n",
    "            dict: A dictionary containing the article title, volume, issue and pages\n",
    "        \"\"\"\n",
    "\n",
    "        article = xml_tree.find(\".//MedlineCitation/Article\")\n",
    "        article_info = {}\n",
    "\n",
    "        # Get the article title\n",
    "        article_info[\"title\"] = \"\"\n",
    "        if article.find(\"ArticleTitle\") is not None:\n",
    "            article_info[\"title\"] = (\n",
    "                self.__stringify_children(article.find(\"ArticleTitle\")) or \"\"\n",
    "            )\n",
    "\n",
    "        # Get the article language\n",
    "        article_info[\"languages\"] = \"\"\n",
    "        if article.find(\"Language\") is not None:\n",
    "            article_info[\"languages\"] = \";\".join(\n",
    "                [language.text for language in article.findall(\"Language\")]\n",
    "            )\n",
    "\n",
    "        # Get the article volume\n",
    "        volume = article.find(\"Journal/JournalIssue/Volume\")\n",
    "        article_info[\"volume\"] = \"\"\n",
    "        if volume is not None:\n",
    "            article_info[\"volume\"] = volume.text or \"\"\n",
    "\n",
    "        # Get the article issue\n",
    "        issue = article.find(\"Journal/JournalIssue/Issue\")\n",
    "        article_info[\"issue\"] = \"\"\n",
    "        if issue is not None:\n",
    "            article_info[\"issue\"] = issue.text or \"\"\n",
    "\n",
    "        # Get the article pages\n",
    "        pages = article.find(\"Pagination/MedlinePgn\")\n",
    "        article_info[\"pages\"] = \"\"\n",
    "        if pages is not None:\n",
    "            article_info[\"pages\"] = pages.text or \"\"\n",
    "\n",
    "        return article_info\n",
    "\n",
    "    def get_authors(self, xml_tree):\n",
    "        \"\"\"Get the authors from the XML tree\n",
    "        Args:\n",
    "            xml_tree (Element): The lxml node pointing to a medline document\n",
    "        Returns:\n",
    "            list: A list of dictionaries containing the first name, middle name, \n",
    "                  last name, suffix, initials, affiliation and email\n",
    "        \"\"\"\n",
    "\n",
    "        authors = []\n",
    "        author_list = xml_tree.findall(\".//AuthorList/Author\")\n",
    "        if author_list is not None:\n",
    "            for author in author_list:\n",
    "                authors.append(\n",
    "                    {\n",
    "                        \"first\": author.find(\"ForeName\").text\n",
    "                        if author.find(\"ForeName\") is not None\n",
    "                        else \"\",\n",
    "                        \"middle\": \"\",\n",
    "                        \"last\": author.find(\"LastName\").text\n",
    "                        if author.find(\"LastName\") is not None\n",
    "                        else \"\",\n",
    "                        \"suffix\": \"\",\n",
    "                        \"initials\": author.find(\"Initials\").text\n",
    "                        if author.find(\"Initials\") is not None\n",
    "                        else \"\",\n",
    "                        \"affiliation\": self.__stringify_children(\n",
    "                            author.find(\"AffiliationInfo/Affiliation\")\n",
    "                        )\n",
    "                        if author.find(\"AffiliationInfo/Affiliation\") is not None\n",
    "                        else \"\",\n",
    "                        \"email\": \"\",\n",
    "                    }\n",
    "                )\n",
    "        return authors\n",
    "\n",
    "    def parse_mesh_terms_with_subs(self, xml_tree):\n",
    "        \"\"\"\n",
    "        Parse the mesh terms with subheadings from the XML tree\n",
    "        Args:\n",
    "            xml_tree (Element): The lxml node pointing to a medline document\n",
    "        Returns:\n",
    "            str: A string of semicolon ``;`` separated mesh terms with subheadings\n",
    "        \"\"\"\n",
    "\n",
    "        mesh = xml_tree.find(\".//MedlineCitation/MeshHeadingList\")\n",
    "        if mesh is not None:\n",
    "            mesh_terms_list = []\n",
    "            for m in mesh.getchildren():\n",
    "                descriptor_name = m.find(\"DescriptorName\")\n",
    "                term = descriptor_name.attrib.get(\"UI\", \"\") + \":\" + descriptor_name.text\n",
    "                if descriptor_name.attrib.get(\"MajorTopicYN\", \"\") == \"Y\":\n",
    "                    term += \"*\"\n",
    "                for q in m.findall(\"QualifierName\"):\n",
    "                    term += \" / \" + q.attrib.get(\"UI\", \"\") + \":\" + q.text\n",
    "                    if q.attrib.get(\"MajorTopicYN\", \"\") == \"Y\":\n",
    "                        term += \"*\"\n",
    "                mesh_terms_list.append(term)\n",
    "            mesh_terms = \"; \".join(mesh_terms_list)\n",
    "        else:\n",
    "            mesh_terms = \"\"\n",
    "        return mesh_terms\n",
    "\n",
    "    def get_article_citation_count(self, doi=None):\n",
    "        \"\"\"Get the citation count of an article based on the DOI using the crossref website\n",
    "        https://api.crossref.org/works/{doi}\n",
    "\n",
    "        Args:\n",
    "            doi (str): The DOI of the article\n",
    "        Returns:\n",
    "            int: The citation count of the article\n",
    "        \"\"\"\n",
    "\n",
    "        if doi in [\"\", None]:\n",
    "            print(\"DOI is empty, cannot get citation count\")\n",
    "            return None\n",
    "        params = {\"mailto\": \"halil@johnsnowlabs.com\"}\n",
    "        url = f\"https://api.crossref.org/works/{doi}\"\n",
    "        response = requests.get(url, params=params, timeout=15)\n",
    "        data = response.json()\n",
    "\n",
    "        if (response.status_code == 200) and (\n",
    "            \"is-referenced-by-count\" in data[\"message\"]\n",
    "        ):\n",
    "            return data[\"message\"][\"is-referenced-by-count\"]\n",
    "\n",
    "        return None\n",
    "\n",
    "    def get_journal_ranking(self, journal_name=None):\n",
    "        \"\"\"Get the ranking of a journal based on the journal name using the exaly website\n",
    "        https://exaly.com/journals/if/1?q=\n",
    "\n",
    "        Args:\n",
    "            journal_name (str): The full name of the journal\n",
    "        Returns:\n",
    "            dict: A dictionary containing the journal ranking info\n",
    "        \"\"\"\n",
    "\n",
    "        if journal_name in [\"\", None]:\n",
    "            print(\"Journal name is empty, cannot get journal ranking\")\n",
    "            return {}\n",
    "\n",
    "        if journal_name in self.journal_ranking_dict:\n",
    "            return self.journal_ranking_dict[journal_name]\n",
    "\n",
    "        query = quote(journal_name.replace(\"&\", \"and\"))\n",
    "\n",
    "        # URL of the webpage containing the table\n",
    "        url = f\"https://exaly.com/journals/if/1?q={query}\"\n",
    "\n",
    "        # Send a GET request to the webpage\n",
    "        response = requests.get(url, timeout=15)\n",
    "\n",
    "        if response.status_code == 200:\n",
    "            # Create a BeautifulSoup object to parse the HTML content\n",
    "            soup = BeautifulSoup(response.content, \"html.parser\")\n",
    "\n",
    "            # Find the table element on the webpage\n",
    "            table = soup.find(\"table\")\n",
    "            ranking = None\n",
    "\n",
    "            if table:\n",
    "                # Extract the data from the table\n",
    "                table_data = []\n",
    "                for row in table.find_all(\"tr\"):\n",
    "                    row_data = []\n",
    "                    for cell in row.find_all([\"th\", \"td\"]):\n",
    "                        row_data.append(cell.get_text(strip=True))\n",
    "                    if row_data:\n",
    "                        table_data.append(row_data)\n",
    "\n",
    "                for row in table_data:\n",
    "                    if (\n",
    "                        row[0].replace(\"&\", \"and\").lower()\n",
    "                        == journal_name.replace(\"&\", \"and\").lower()\n",
    "                    ):\n",
    "                        ranking = dict(zip(table_data[0], row))\n",
    "                        break\n",
    "\n",
    "                if ranking is None:\n",
    "                    # Create a dictionary from the first two rows of the table\n",
    "                    ranking = dict(zip(table_data[0], table_data[1]))\n",
    "\n",
    "                _ = ranking.pop(\"star\")  # remove the star column\n",
    "\n",
    "                # add the journal name to the dictionary\n",
    "                self.journal_ranking_dict[journal_name] = ranking\n",
    "            else:\n",
    "                ranking = {}\n",
    "        else:\n",
    "            ranking = {}\n",
    "\n",
    "        return ranking\n",
    "\n",
    "    def get_sentences(self, content):\n",
    "        \"\"\"Get the sentences from the content and return them as a list\"\"\"\n",
    "\n",
    "        formater = Formater()\n",
    "        _UNPROTECTED_BREAK_INDICATOR = \"(?![^]*)\"\n",
    "\n",
    "        content_modified = formater.generate_protecedContent(content)\n",
    "\n",
    "        list_sentences = re.split(_UNPROTECTED_BREAK_INDICATOR, content_modified)\n",
    "\n",
    "        list_sentences = [\n",
    "            sentence.replace(BREAK_INDICATOR, \"\") for sentence in list_sentences\n",
    "        ]\n",
    "        list_sentences = [\n",
    "            sentence.replace(BREAK_INDICATOR, \"\") for sentence in list_sentences\n",
    "        ]\n",
    "\n",
    "        def recover_senctence(sentence: str):\n",
    "            wip_sentence = sentence\n",
    "            for key, item in reverse_dict.items():\n",
    "                wip_sentence = wip_sentence.replace(key, item)\n",
    "            return wip_sentence\n",
    "\n",
    "        list_sentences = [recover_senctence(sentence) for sentence in list_sentences]\n",
    "        list_sentences = [\n",
    "            sentence.strip() for sentence in list_sentences if sentence.strip() != \"\"\n",
    "        ]\n",
    "\n",
    "        return list_sentences\n",
    "\n",
    "    def split_article_paragraphs(self, json_data):\n",
    "        \"\"\"Split the article into paragraphs of max_token_length & min_token_length tokens\n",
    "        with an overlap of sentence_overlap sentences\"\"\"\n",
    "\n",
    "        def group_sentences(sentences, max_tokens, overlap_length):\n",
    "            \"\"\"Group sentences into groups of max_tokens tokens \n",
    "            with an overlap of overlap_length sentences\"\"\"\n",
    "            grouped_sentences = []\n",
    "            current_group = []\n",
    "            grouped_sentences_lengths = []\n",
    "            token_count = 0\n",
    "            tokenizer = tiktoken.get_encoding(\"p50k_base\")\n",
    "\n",
    "            for sentence in sentences:\n",
    "                tokens = tokenizer.encode(sentence, disallowed_special=())\n",
    "                sentence_length = len(tokens)\n",
    "\n",
    "                if token_count + sentence_length <= max_tokens:\n",
    "                    current_group.append(sentence)\n",
    "                    token_count += sentence_length\n",
    "                else:\n",
    "                    grouped_sentences.append(current_group)\n",
    "                    grouped_sentences_lengths.append(token_count)\n",
    "                    current_group = current_group[\n",
    "                        -overlap_length:\n",
    "                    ]  # Start next group with the last overlap_length sentences\n",
    "                    current_group.append(sentence)\n",
    "                    token_count = sum(\n",
    "                        len(tokenizer.encode(sent, disallowed_special=()))\n",
    "                        for sent in current_group\n",
    "                    )\n",
    "\n",
    "            # Add the last group if it is not empty\n",
    "            if current_group:\n",
    "                grouped_sentences.append(current_group)\n",
    "                grouped_sentences_lengths.append(token_count)\n",
    "\n",
    "            return grouped_sentences, grouped_sentences_lengths\n",
    "\n",
    "        abstract_text = json_data[\"abstract\"][0].get(\"text\", \"\")\n",
    "        pmid = json_data[\"pmid\"]\n",
    "\n",
    "        if abstract_text != \"\" and pmid != \"\":\n",
    "            sentences = self.get_sentences(abstract_text)\n",
    "        else:\n",
    "            return None\n",
    "\n",
    "        article_splits = []\n",
    "\n",
    "        # split the abstract\n",
    "        if json_data[\"abstract\"][0] is not None:\n",
    "            split_id = 0\n",
    "            splits, splits_lengths = group_sentences(\n",
    "                sentences, self.max_split_token_length, self.sentence_overlap\n",
    "            )\n",
    "\n",
    "            #  add the splits to the splits list with the corresponding split_id and text\n",
    "            split_list = []\n",
    "            for i, split in enumerate(splits):\n",
    "                \n",
    "                if (splits_lengths[i] > self.min_split_token_length) or (i == 0):\n",
    "                    split_list.append(\n",
    "                        {f\"PMID{pmid}_abstract_split_{split_id}\": \" \".join(split)}\n",
    "                    )\n",
    "                    split_id += 1\n",
    "                else:\n",
    "                    # merge the split with previous split if it is less than min_split_token_length\n",
    "                    # trim the split overlap_length sentences\n",
    "                    previous_split = split_list[-1]\n",
    "                    split_sentences = self.get_sentences(\" \".join(split))\n",
    "                    previous_split[list(previous_split.keys())[0]] += \" \".join(split_sentences[self.sentence_overlap :])\n",
    "            article_splits.extend(split_list)\n",
    "\n",
    "        return article_splits\n",
    "\n",
    "    def build_pubmed_json(self, xml_path=None, xml_tree=None):\n",
    "        \"\"\"Build the JSON object for the article\n",
    "        Args:\n",
    "            xml_path (str): The path to the XML file of the article\n",
    "            xml_tree (Element): The lxml node pointing to a medline document\n",
    "        Returns:\n",
    "            dict: A dictionary containing the JSON object for the article\n",
    "        \"\"\"\n",
    "\n",
    "        # Get the XML tree for the article if path is provided\n",
    "        if xml_path is not None:\n",
    "            xml_tree = etree.parse(xml_path)\n",
    "        # Get the abstract text\n",
    "        abstract = self.get_abstract(xml_tree)\n",
    "\n",
    "        # Return None if the abstract is empty\n",
    "        if abstract is None:\n",
    "            return None\n",
    "\n",
    "        # Build the JSON object\n",
    "        json_output = {\n",
    "            \"abstract\": [abstract],\n",
    "            \"meta_info\": {\n",
    "                \"articleids\": self.get_article_ids(xml_tree),\n",
    "                \"kwd\": self.get_keywords(xml_tree),\n",
    "                \"dates_history\": self.get_dates_history(xml_tree),\n",
    "                **self.get_journal_info(xml_tree),\n",
    "                **self.get_article_info(xml_tree),\n",
    "                \"authors\": self.get_authors(xml_tree),\n",
    "                \"mesh_terms\": self.parse_mesh_terms_with_subs(xml_tree),\n",
    "            },\n",
    "        }\n",
    "\n",
    "        journal_name = json_output[\"meta_info\"].get(\"fulljournalname\", \"\")\n",
    "\n",
    "        doi = \"\"\n",
    "        pmid = \"\"\n",
    "        for id_item in json_output[\"meta_info\"].get(\"articleids\", []):\n",
    "            if id_item[\"idtype\"] == \"doi\":\n",
    "                doi = id_item[\"value\"]\n",
    "            elif id_item[\"idtype\"] == \"pubmed\":\n",
    "                pmid = id_item[\"value\"]\n",
    "\n",
    "        json_output[\"pmid\"] = pmid\n",
    "\n",
    "        # Get the article splits\n",
    "        json_output[\"article_splits\"] = self.split_article_paragraphs(json_output)\n",
    "\n",
    "        # Get the citation count of the article\n",
    "        if self.citation_count_bool:\n",
    "            json_output[\"meta_info\"][\n",
    "                \"citation_count\"\n",
    "            ] = self.get_article_citation_count(doi)\n",
    "\n",
    "        # # Get the journal ranking info\n",
    "        if self.journal_ranking_bool:\n",
    "            json_output[\"meta_info\"][\"journal_ranking\"] = self.get_journal_ranking(\n",
    "                journal_name\n",
    "            )\n",
    "\n",
    "        return json_output\n",
    "\n",
    "    def parse_pubmed_xml_iter(self, path):\n",
    "        \"\"\"Parse the XML file of the article and yield the JSON object for the article\n",
    "        Args:\n",
    "            path (str): The path to the XML GZ file of the article\n",
    "        Yields:\n",
    "            dict: A dictionary containing the JSON object for the article\n",
    "        \"\"\"\n",
    "        with gzip.open(path, \"rb\") as f:\n",
    "            for _, element in etree.iterparse(f, events=(\"end\",)):\n",
    "                if element.tag == \"PubmedArticle\":\n",
    "                    res = self.build_pubmed_json(xml_tree=element)\n",
    "                    element.clear()\n",
    "                    yield res"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Process a large xml gz file that contains more than around 5000 articles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "pubmed = PubmedPMID()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "\n",
    "dicts_out = pubmed.parse_pubmed_xml_iter(\"data/pubmed23n1166.xml.gz\")\n",
    "\n",
    "i = 0\n",
    "for article in dicts_out:\n",
    "    if article is None:\n",
    "        continue\n",
    "    pmid = article[\"pmid\"]\n",
    "    with open(f\"./data/pubmed/pubmed_{pmid}.json\", \"w\") as f:\n",
    "        json.dump(article, f)\n",
    "    # i += 1\n",
    "    # if i > 20:\n",
    "    #     break"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Fetch the xml file from pubmed API"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'<?xml version=\"1.0\" ?>\\n<!DOCTYPE PubmedArticleSet PUBLIC \"-//NLM//DTD PubMedArticle, 1st January 2023//EN\" \"https://dtd.nlm.nih.gov/ncbi/pubmed/out/pubmed_230101.dtd\">\\n<PubmedArticleSet>\\n<PubmedArticle><MedlineCitation Status=\"PubMed-not-MEDLINE\" Owner=\"NLM\"><PMID Version=\"1\">36464825</PMID><DateRevised><Year>2022</Year><Month>12</Month><Day>14</Day></DateRevised><Article PubModel=\"Print-Electronic\"><Journal><ISSN IssnType=\"Print\">2234-2400</ISSN><JournalIssue CitedMedium=\"Print\"><Volume>55</Volume><Issue>6</Issue><PubDate><Year>2022</Year><Month>Nov</Month></PubDate></JournalIssue><Title>Clinical endoscopy</Title><ISOAbbreviation>Clin Endosc</ISOAbbreviation></Journal><ArticleTitle>Underestimation of endoscopic size in large gastric epithelial neoplasms.</ArticleTitle><Pagination><StartPage>760</StartPage><EndPage>766</EndPage><MedlinePgn>760-766</MedlinePgn></Pagination><ELocationID EIdType=\"doi\" ValidYN=\"Y\">10.5946/ce.2021.269</ELocationID><Abstract><AbstractText Label=\"BACKGROUND/AIMS\" NlmCategory=\"OBJECTIVE\">Endoscopic submucosal dissection (ESD) is an effective method for resecting gastric adenomas and adenocarcinomas. A significant discrepancy was observed between endoscopic and pathological sizes in samples obtained from patients undergoing ESD. This study elucidates the factors affecting size discrepancy after formalin fixation.</AbstractText><AbstractText Label=\"METHODS\" NlmCategory=\"METHODS\">The records of 64 patients with 69 lesions were analyzed, including 50 adenomas and 19 adenocarcinomas. Data on location, gross shape, histology, and size after fixation in formalin were collected.</AbstractText><AbstractText Label=\"RESULTS\" NlmCategory=\"RESULTS\">The mean size of the resected specimen appeared to decrease after formalin fixation (37.5 mm prefixation vs. 35.8 mm postfixation, p&lt;0.05). The mean long axis diameter of the lesions was 20.3&#xb1;7.9 mm prefixation and 13.4&#xb1;7.9 mm postfixation. Size differences in lesions smaller than 20 mm were significantly greater than those in lesions larger than 20 mm (7.6&#xb1;5.6 mm vs. 2.5&#xb1;5.8 mm, p&lt;0.01). In multivariate analysis, a tumor size of &#x2265;20 mm was found to be an independent factor affecting size postformalin fixation (p&lt;0.05).</AbstractText><AbstractText Label=\"CONCLUSION\" NlmCategory=\"CONCLUSIONS\">The endoscopic size of lesions before ESD may be underestimated in tumors larger than 20 mm in size. Therefore, increased attention must be paid during ESD to avoid instances of incomplete resection.</AbstractText></Abstract><AuthorList CompleteYN=\"Y\"><Author ValidYN=\"Y\"><LastName>Song</LastName><ForeName>Jae Sun</ForeName><Initials>JS</Initials><AffiliationInfo><Affiliation>Division of Gastroenterology, Department of Internal Medicine, Presbyterian Medical Center, Jeonju, Korea.</Affiliation></AffiliationInfo></Author><Author ValidYN=\"Y\"><LastName>Kim</LastName><ForeName>Byung Sun</ForeName><Initials>BS</Initials><AffiliationInfo><Affiliation>Division of Gastroenterology, Department of Internal Medicine, Presbyterian Medical Center, Jeonju, Korea.</Affiliation></AffiliationInfo></Author><Author ValidYN=\"Y\"><LastName>Yang</LastName><ForeName>Min A</ForeName><Initials>MA</Initials><AffiliationInfo><Affiliation>Division of Gastroenterology, Department of Internal Medicine, Presbyterian Medical Center, Jeonju, Korea.</Affiliation></AffiliationInfo></Author><Author ValidYN=\"Y\"><LastName>Lee</LastName><ForeName>Young Jae</ForeName><Initials>YJ</Initials><AffiliationInfo><Affiliation>Division of Gastroenterology, Department of Internal Medicine, Presbyterian Medical Center, Jeonju, Korea.</Affiliation></AffiliationInfo></Author><Author ValidYN=\"Y\"><LastName>Jung</LastName><ForeName>Gum Mo</ForeName><Initials>GM</Initials><AffiliationInfo><Affiliation>Division of Gastroenterology, Department of Internal Medicine, Presbyterian Medical Center, Jeonju, Korea.</Affiliation></AffiliationInfo></Author><Author ValidYN=\"Y\"><LastName>Kim</LastName><ForeName>Ji Woong</ForeName><Initials>JW</Initials><AffiliationInfo><Affiliation>Division of Gastroenterology, Department of Internal Medicine, Presbyterian Medical Center, Jeonju, Korea.</Affiliation></AffiliationInfo></Author><Author ValidYN=\"Y\"><LastName>Cho</LastName><ForeName>Jin Woong</ForeName><Initials>JW</Initials><AffiliationInfo><Affiliation>Division of Gastroenterology, Department of Internal Medicine, Presbyterian Medical Center, Jeonju, Korea.</Affiliation></AffiliationInfo></Author></AuthorList><Language>eng</Language><PublicationTypeList><PublicationType UI=\"D016428\">Journal Article</PublicationType></PublicationTypeList><ArticleDate DateType=\"Electronic\"><Year>2022</Year><Month>05</Month><Day>19</Day></ArticleDate></Article><MedlineJournalInfo><Country>Korea (South)</Country><MedlineTA>Clin Endosc</MedlineTA><NlmUniqueID>101576886</NlmUniqueID><ISSNLinking>2234-2400</ISSNLinking></MedlineJournalInfo><CommentsCorrectionsList><CommentsCorrections RefType=\"CommentIn\"><RefSource>Clin Endosc. 2022 Nov;55(6):753-754</RefSource><PMID Version=\"1\">36464822</PMID></CommentsCorrections></CommentsCorrectionsList><KeywordList Owner=\"NOTNLM\"><Keyword MajorTopicYN=\"N\">Endoscopic submucosal dissection</Keyword><Keyword MajorTopicYN=\"N\">Endoscopy</Keyword><Keyword MajorTopicYN=\"N\">Stomach neoplasms</Keyword></KeywordList><CoiStatement>\\n<b>Conflicts of Interest</b>\\n. The authors have no potential conflicts of interest.</CoiStatement></MedlineCitation><PubmedData><History><PubMedPubDate PubStatus=\"received\"><Year>2021</Year><Month>11</Month><Day>3</Day></PubMedPubDate><PubMedPubDate PubStatus=\"accepted\"><Year>2022</Year><Month>2</Month><Day>9</Day></PubMedPubDate><PubMedPubDate PubStatus=\"entrez\"><Year>2022</Year><Month>12</Month><Day>5</Day><Hour>0</Hour><Minute>42</Minute></PubMedPubDate><PubMedPubDate PubStatus=\"pubmed\"><Year>2022</Year><Month>12</Month><Day>6</Day><Hour>6</Hour><Minute>0</Minute></PubMedPubDate><PubMedPubDate PubStatus=\"medline\"><Year>2022</Year><Month>12</Month><Day>6</Day><Hour>6</Hour><Minute>1</Minute></PubMedPubDate></History><PublicationStatus>ppublish</PublicationStatus><ArticleIdList><ArticleId IdType=\"pubmed\">36464825</ArticleId><ArticleId IdType=\"pmc\">PMC9726444</ArticleId><ArticleId IdType=\"doi\">10.5946/ce.2021.269</ArticleId><ArticleId IdType=\"pii\">ce.2021.269</ArticleId></ArticleIdList><ReferenceList><Reference><Citation>Chung IK, Lee JH, Lee SH, et al. Therapeutic outcomes in 1000 cases of endoscopic submucosal dissection for early gastric neoplasms: Korean ESD Study Group multicenter study. Gastrointest Endosc. 2009;69:1228&#x2013;1235.</Citation><ArticleIdList><ArticleId IdType=\"pubmed\">19249769</ArticleId></ArticleIdList></Reference><Reference><Citation>Min BH, Kim ER, Kim KM, et al. Surveillance strategy based on the incidence and patterns of recurrence after curative endoscopic submucosal dissection for early gastric cancer. Endoscopy. 2015;47:784&#x2013;793.</Citation><ArticleIdList><ArticleId IdType=\"pubmed\">26111362</ArticleId></ArticleIdList></Reference><Reference><Citation>Siu KF, Cheung HC, Wong J. Shrinkage of the esophagus after resection for carcinoma. Ann Surg. 1986;203:173&#x2013;176.</Citation><ArticleIdList><ArticleId IdType=\"pmc\">PMC1251066</ArticleId><ArticleId IdType=\"pubmed\">3947154</ArticleId></ArticleIdList></Reference><Reference><Citation>Goldstein NS, Soman A, Sacksner J. Disparate surgical margin lengths of colorectal resection specimens between in vivo and in vitro measurements: the effects of surgical resection and formalin fixation on organ shrinkage. Am J Clin Pathol. 1999;111:349&#x2013;351.</Citation><ArticleIdList><ArticleId IdType=\"pubmed\">10078110</ArticleId></ArticleIdList></Reference><Reference><Citation>Kim EH, Park JC, Song IJ, et al. Prediction model for non-curative resection of endoscopic submucosal dissection in patients with early gastric cancer. Gastrointest Endosc. 2017;85:976&#x2013;983.</Citation><ArticleIdList><ArticleId IdType=\"pubmed\">27756614</ArticleId></ArticleIdList></Reference><Reference><Citation>Gotoda T, Yanagisawa A, Sasako M, et al. Incidence of lymph node metastasis from early gastric cancer: estimation with a large number of cases at two large centers. Gastric Cancer. 2000;3:219&#x2013;225.</Citation><ArticleIdList><ArticleId IdType=\"pubmed\">11984739</ArticleId></ArticleIdList></Reference><Reference><Citation>Cho JW. The role of endosonography in the staging of gastrointestinal cancers. Clin Endosc. 2015;48:297&#x2013;301.</Citation><ArticleIdList><ArticleId IdType=\"pmc\">PMC4522420</ArticleId><ArticleId IdType=\"pubmed\">26240802</ArticleId></ArticleIdList></Reference><Reference><Citation>Shim CN, Song MK, Kang DR, et al. Size discrepancy between endoscopic size and pathologic size is not negligible in endoscopic resection for early gastric cancer. Surg Endosc. 2014;28:2199&#x2013;2207.</Citation><ArticleIdList><ArticleId IdType=\"pubmed\">24519032</ArticleId></ArticleIdList></Reference><Reference><Citation>Asada-Hirayama I, Kodashima S, Goto O, et al. Factors predictive of inaccurate determination of horizontal extent of intestinal-type early gastric cancers during endoscopic submucosal dissection: a retrospective analysis. Dig Endosc. 2013;25:593&#x2013;600.</Citation><ArticleIdList><ArticleId IdType=\"pubmed\">23489945</ArticleId></ArticleIdList></Reference><Reference><Citation>Sohn YJ, Jang JS, Choi SR, et al. Early detection of recurrence after endoscopic treatment for early gastric cancer. Scand J Gastroenterol. 2009;44:1109&#x2013;1114.</Citation><ArticleIdList><ArticleId IdType=\"pubmed\">19593687</ArticleId></ArticleIdList></Reference><Reference><Citation>Yamamoto Y, Fujisaki J, Hirasawa T, et al. Therapeutic outcomes of endoscopic submucosal dissection of undifferentiated-type intramucosal gastric cancer without ulceration and preoperatively diagnosed as 20 millimetres or less in diameter. Dig Endosc. 2010;22:112&#x2013;118.</Citation><ArticleIdList><ArticleId IdType=\"pubmed\">20447204</ArticleId></ArticleIdList></Reference><Reference><Citation>Schoen RE, Gerber LD, Margulies C. The pathologic measurement of polyp size is preferable to the endoscopic estimate. Gastrointest Endosc. 1997;46:492&#x2013;496.</Citation><ArticleIdList><ArticleId IdType=\"pubmed\">9434214</ArticleId></ArticleIdList></Reference><Reference><Citation>Fennerty MB, Davidson J, Emerson SS, et al. Are endoscopic measurements of colonic polyps reliable? Am J Gastroenterol. 1993;88:496&#x2013;500.</Citation><ArticleIdList><ArticleId IdType=\"pubmed\">8470627</ArticleId></ArticleIdList></Reference><Reference><Citation>Gopalswamy N, Shenoy VN, Choudhry U, et al. Is in vivo measurement of size of polyps during colonoscopy accurate? Gastrointest Endosc. 1997;46:497&#x2013;502.</Citation><ArticleIdList><ArticleId IdType=\"pubmed\">9434215</ArticleId></ArticleIdList></Reference><Reference><Citation>Yun GW, Kim JH, Lee YC, et al. What are the risk factors for residual tumor cells after endoscopic complete resection in gastric epithelial neoplasia? Surg Endosc. 2015;29:487&#x2013;492.</Citation><ArticleIdList><ArticleId IdType=\"pubmed\">25015521</ArticleId></ArticleIdList></Reference><Reference><Citation>Kim SB, Lee SH, Bae SI, et al. Association between Helicobacter pylori status and metachronous gastric cancer after endoscopic resection. World J Gastroenterol. 2016;22:9794&#x2013;9802.</Citation><ArticleIdList><ArticleId IdType=\"pmc\">PMC5124984</ArticleId><ArticleId IdType=\"pubmed\">27956803</ArticleId></ArticleIdList></Reference></ReferenceList></PubmedData></PubmedArticle></PubmedArticleSet>'"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pubmed.get_pubmed_article_xml(\"36464825\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Process a single xml file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'abstract': [{'text': '\\n\\nBackground:\\nAmyotrophic lateral sclerosis (ALS) is a progressive neurodegenerative disease caused by the loss of motor neurons, and development of effective medicines is urgently required. Induced pluripotent stem cell (iPSC)-based drug repurposing identified the Src/c-Abl inhibitor bosutinib, which is approved for the treatment of chronic myelogenous leukemia (CML), as a candidate for the molecular targeted therapy of ALS.\\n\\n\\nMethods:\\nAn open-label, multicentre, dose-escalation phase 1 study using a 3\\xa0+\\xa03 design was conducted in 4 hospitals in Japan to evaluate the safety and tolerability of bosutinib in patients with ALS. Furthermore, the exploratory efficacy was evaluated using Revised ALS Functional Rating Scale (ALSFRS-R), predictive biomarkers including plasma neurofilament light chain (NFL) were explored, and single-cell RNA sequencing of iPSC-derived motor neurons was conducted. Patients, whose total ALSFRS-R scores decreased by 1-3 points during the 12-week, received escalating doses starting from 100\\xa0mg quaque die (QD) up to 400\\xa0mg QD based on dose-limiting toxicity (DLT) occurrence, and all participants who received one dose of the study drug were included in the primary analysis. This trial is registered with ClinicalTrials.gov, NCT04744532, as Induced pluripotent stem cell-based Drug Repurposing for Amyotrophic Lateral Sclerosis Medicine (iDReAM) study.\\n\\n\\nFindings:\\nBetween March 29, 2019 and May 7, 2021, 20 patients were enrolled, 13 of whom received bosutinib treatment and 12 were included in the safety and efficacy analyses. No DLTs were observed up to 300\\xa0mg QD, but DLTs were observed in 3/3 patients of the 400\\xa0mg QD cohort. In all patients receiving 100\\xa0mg-400\\xa0mg, the prevalent adverse events (AEs) were gastrointestinal AEs in 12 patients (92.3%), liver function related AEs in 7 patients (53.8%), and rash in 3 patients (23.1%). The safety profile was consistent with that known for CML treatment, and ALS-specific AEs were not observed. A subset of patients (5/9 patients) was found to respond well to bosutinib treatment over the 12-week treatment period. It was found that the treatment-responsive patients could be distinguished by their lower levels of plasma NFL. Furthermore, single-cell RNA sequencing of iPSC-derived motor neurons revealed the pathogenesis related molecular signature in patients with ALS showing responsiveness to bosutinib.\\n\\n\\nInterpretation:\\nThis is the first trial of a Src/c-Abl inhibitor, bosutinib, for patients with ALS. The safety and tolerability of bosutinib up to 300\\xa0mg, not 400\\xa0mg, in ALS were described, and responsiveness of patients on motor function was observed. Since this was an open-label trial within a short period with a limited number of patients, further clinical trials will be required.\\n\\n\\nFunding:\\nAMED and iPS Cell Research Fund.',\n",
       "   'section_title': 'Abstract',\n",
       "   'section_type': 'ABSTRACT'}],\n",
       " 'meta_info': {'articleids': [{'idtype': 'pubmed', 'value': '36467452'},\n",
       "   {'idtype': 'doi', 'value': '10.1016/j.eclinm.2022.101707'},\n",
       "   {'idtype': 'pii', 'value': 'S2589-5370(22)00437-0'},\n",
       "   {'idtype': 'pmc', 'value': 'PMC9716331'}],\n",
       "  'kwd': '',\n",
       "  'dates_history': [{'date_type': 'received',\n",
       "    'year': '2022',\n",
       "    'month': '08',\n",
       "    'day': '05'},\n",
       "   {'date_type': 'revised', 'year': '2022', 'month': '09', 'day': '29'},\n",
       "   {'date_type': 'accepted', 'year': '2022', 'month': '09', 'day': '30'},\n",
       "   {'date_type': 'entrez', 'year': '2022', 'month': '12', 'day': '5'},\n",
       "   {'date_type': 'pubmed', 'year': '2022', 'month': '12', 'day': '6'},\n",
       "   {'date_type': 'medline', 'year': '2022', 'month': '12', 'day': '6'}],\n",
       "  'fulljournalname': 'EClinicalMedicine',\n",
       "  'journal_abbrev': 'EClinicalMedicine',\n",
       "  'issn': '2589-5370',\n",
       "  'pubdate': '2022 Nov',\n",
       "  'title': 'Safety and tolerability of bosutinib in patients with amyotrophic lateral sclerosis (iDReAM study): A multicentre, open-label, dose-escalation phase 1 trial.',\n",
       "  'languages': 'eng',\n",
       "  'volume': '53',\n",
       "  'issue': '',\n",
       "  'pages': '101707',\n",
       "  'authors': [{'first': 'Keiko',\n",
       "    'middle': '',\n",
       "    'last': 'Imamura',\n",
       "    'suffix': '',\n",
       "    'initials': 'K',\n",
       "    'affiliation': 'Center for iPS Cell Research and Application (CiRA), Kyoto University, Kyoto, Japan.',\n",
       "    'email': ''},\n",
       "   {'first': 'Yuishin',\n",
       "    'middle': '',\n",
       "    'last': 'Izumi',\n",
       "    'suffix': '',\n",
       "    'initials': 'Y',\n",
       "    'affiliation': 'Department of Neurology, Tokushima University Graduate School of Biomedical Sciences, Tokushima, Japan.',\n",
       "    'email': ''},\n",
       "   {'first': 'Makiko',\n",
       "    'middle': '',\n",
       "    'last': 'Nagai',\n",
       "    'suffix': '',\n",
       "    'initials': 'M',\n",
       "    'affiliation': 'Department of Neurology, Kitasato University School of Medicine, Sagamihara, Japan.',\n",
       "    'email': ''},\n",
       "   {'first': 'Kazutoshi',\n",
       "    'middle': '',\n",
       "    'last': 'Nishiyama',\n",
       "    'suffix': '',\n",
       "    'initials': 'K',\n",
       "    'affiliation': 'Department of Neurology, Kitasato University School of Medicine, Sagamihara, Japan.',\n",
       "    'email': ''},\n",
       "   {'first': 'Yasuhiro',\n",
       "    'middle': '',\n",
       "    'last': 'Watanabe',\n",
       "    'suffix': '',\n",
       "    'initials': 'Y',\n",
       "    'affiliation': 'Division of Neurology, Department of Brain and Neurosciences, Faculty of Medicine, Tottori University, Yonago, Japan.',\n",
       "    'email': ''},\n",
       "   {'first': 'Ritsuko',\n",
       "    'middle': '',\n",
       "    'last': 'Hanajima',\n",
       "    'suffix': '',\n",
       "    'initials': 'R',\n",
       "    'affiliation': 'Division of Neurology, Department of Brain and Neurosciences, Faculty of Medicine, Tottori University, Yonago, Japan.',\n",
       "    'email': ''},\n",
       "   {'first': 'Naohiro',\n",
       "    'middle': '',\n",
       "    'last': 'Egawa',\n",
       "    'suffix': '',\n",
       "    'initials': 'N',\n",
       "    'affiliation': 'Department of Neurology, Graduate School of Medicine, Kyoto University, Kyoto, Japan.',\n",
       "    'email': ''},\n",
       "   {'first': 'Takashi',\n",
       "    'middle': '',\n",
       "    'last': 'Ayaki',\n",
       "    'suffix': '',\n",
       "    'initials': 'T',\n",
       "    'affiliation': 'Department of Neurology, Graduate School of Medicine, Kyoto University, Kyoto, Japan.',\n",
       "    'email': ''},\n",
       "   {'first': 'Ryosuke',\n",
       "    'middle': '',\n",
       "    'last': 'Oki',\n",
       "    'suffix': '',\n",
       "    'initials': 'R',\n",
       "    'affiliation': 'Department of Neurology, Tokushima University Graduate School of Biomedical Sciences, Tokushima, Japan.',\n",
       "    'email': ''},\n",
       "   {'first': 'Koji',\n",
       "    'middle': '',\n",
       "    'last': 'Fujita',\n",
       "    'suffix': '',\n",
       "    'initials': 'K',\n",
       "    'affiliation': 'Department of Neurology, Tokushima University Graduate School of Biomedical Sciences, Tokushima, Japan.',\n",
       "    'email': ''},\n",
       "   {'first': 'Ryuji',\n",
       "    'middle': '',\n",
       "    'last': 'Uozumi',\n",
       "    'suffix': '',\n",
       "    'initials': 'R',\n",
       "    'affiliation': 'Department of Biomedical Statistics and Bioinformatics, Kyoto University, Kyoto, Japan.',\n",
       "    'email': ''},\n",
       "   {'first': 'Akiko',\n",
       "    'middle': '',\n",
       "    'last': 'Morinaga',\n",
       "    'suffix': '',\n",
       "    'initials': 'A',\n",
       "    'affiliation': 'Pfizer R&D Japan G.K., Tokyo, Japan.',\n",
       "    'email': ''},\n",
       "   {'first': 'Tomoko',\n",
       "    'middle': '',\n",
       "    'last': 'Hirohashi',\n",
       "    'suffix': '',\n",
       "    'initials': 'T',\n",
       "    'affiliation': 'Pfizer Inc., New York, New York, USA.',\n",
       "    'email': ''},\n",
       "   {'first': 'Yosuke',\n",
       "    'middle': '',\n",
       "    'last': 'Fujii',\n",
       "    'suffix': '',\n",
       "    'initials': 'Y',\n",
       "    'affiliation': 'Pfizer R&D Japan G.K., Tokyo, Japan.',\n",
       "    'email': ''},\n",
       "   {'first': 'Takuya',\n",
       "    'middle': '',\n",
       "    'last': 'Yamamoto',\n",
       "    'suffix': '',\n",
       "    'initials': 'T',\n",
       "    'affiliation': 'Center for iPS Cell Research and Application (CiRA), Kyoto University, Kyoto, Japan.',\n",
       "    'email': ''},\n",
       "   {'first': 'Harutsugu',\n",
       "    'middle': '',\n",
       "    'last': 'Tatebe',\n",
       "    'suffix': '',\n",
       "    'initials': 'H',\n",
       "    'affiliation': 'Department of Functional Brain Imaging, Institute for Quantum Medical Science, National Institutes for Quantum Science and Technology, Chiba, Japan.',\n",
       "    'email': ''},\n",
       "   {'first': 'Takahiko',\n",
       "    'middle': '',\n",
       "    'last': 'Tokuda',\n",
       "    'suffix': '',\n",
       "    'initials': 'T',\n",
       "    'affiliation': 'Department of Functional Brain Imaging, Institute for Quantum Medical Science, National Institutes for Quantum Science and Technology, Chiba, Japan.',\n",
       "    'email': ''},\n",
       "   {'first': 'Naoto',\n",
       "    'middle': '',\n",
       "    'last': 'Takahashi',\n",
       "    'suffix': '',\n",
       "    'initials': 'N',\n",
       "    'affiliation': 'Department of Hematology, Nephrology, and Rheumatology, Akita University Graduate School of Medicine, Akita, Japan.',\n",
       "    'email': ''},\n",
       "   {'first': 'Satoshi',\n",
       "    'middle': '',\n",
       "    'last': 'Morita',\n",
       "    'suffix': '',\n",
       "    'initials': 'S',\n",
       "    'affiliation': 'Department of Biomedical Statistics and Bioinformatics, Kyoto University, Kyoto, Japan.',\n",
       "    'email': ''},\n",
       "   {'first': 'Ryosuke',\n",
       "    'middle': '',\n",
       "    'last': 'Takahashi',\n",
       "    'suffix': '',\n",
       "    'initials': 'R',\n",
       "    'affiliation': 'Department of Neurology, Graduate School of Medicine, Kyoto University, Kyoto, Japan.',\n",
       "    'email': ''},\n",
       "   {'first': 'Haruhisa',\n",
       "    'middle': '',\n",
       "    'last': 'Inoue',\n",
       "    'suffix': '',\n",
       "    'initials': 'H',\n",
       "    'affiliation': 'Center for iPS Cell Research and Application (CiRA), Kyoto University, Kyoto, Japan.',\n",
       "    'email': ''}],\n",
       "  'mesh_terms': ''},\n",
       " 'pmid': '36467452',\n",
       " 'article_splits': [{'PMID36467452_abstract_split_0': 'Background:\\nAmyotrophic lateral sclerosis (ALS) is a progressive neurodegenerative disease caused by the loss of motor neurons, and development of effective medicines is urgently required. Induced pluripotent stem cell (iPSC)-based drug repurposing identified the Src/c-Abl inhibitor bosutinib, which is approved for the treatment of chronic myelogenous leukemia (CML), as a candidate for the molecular targeted therapy of ALS. Methods:\\nAn open-label, multicentre, dose-escalation phase 1 study using a 3\\xa0+\\xa03 design was conducted in 4 hospitals in Japan to evaluate the safety and tolerability of bosutinib in patients with ALS. Furthermore, the exploratory efficacy was evaluated using Revised ALS Functional Rating Scale (ALSFRS-R), predictive biomarkers including plasma neurofilament light chain (NFL) were explored, and single-cell RNA sequencing of iPSC-derived motor neurons was conducted. Patients, whose total ALSFRS-R scores decreased by 1-3 points during the 12-week, received escalating doses starting from 100\\xa0mg quaque die (QD) up to 400\\xa0mg QD based on dose-limiting toxicity (DLT) occurrence, and all participants who received one dose of the study drug were included in the primary analysis. This trial is registered with ClinicalTrials.gov, NCT04744532, as Induced pluripotent stem cell-based Drug Repurposing for Amyotrophic Lateral Sclerosis Medicine (iDReAM) study. Findings:\\nBetween March 29, 2019 and May 7, 2021, 20 patients were enrolled, 13 of whom received bosutinib treatment and 12 were included in the safety and efficacy analyses. No DLTs were observed up to 300\\xa0mg QD, but DLTs were observed in 3/3 patients of the 400\\xa0mg QD cohort. In all patients receiving 100\\xa0mg-400\\xa0mg, the prevalent adverse events (AEs) were gastrointestinal AEs in 12 patients (92.3%), liver function related AEs in 7 patients (53.8%), and rash in 3 patients (23.1%). The safety profile was consistent with that known for CML treatment, and ALS-specific AEs were not observed.'},\n",
       "  {'PMID36467452_abstract_split_1': 'In all patients receiving 100\\xa0mg-400\\xa0mg, the prevalent adverse events (AEs) were gastrointestinal AEs in 12 patients (92.3%), liver function related AEs in 7 patients (53.8%), and rash in 3 patients (23.1%). The safety profile was consistent with that known for CML treatment, and ALS-specific AEs were not observed. A subset of patients (5/9 patients) was found to respond well to bosutinib treatment over the 12-week treatment period. It was found that the treatment-responsive patients could be distinguished by their lower levels of plasma NFL. Furthermore, single-cell RNA sequencing of iPSC-derived motor neurons revealed the pathogenesis related molecular signature in patients with ALS showing responsiveness to bosutinib. Interpretation:\\nThis is the first trial of a Src/c-Abl inhibitor, bosutinib, for patients with ALS. The safety and tolerability of bosutinib up to 300\\xa0mg, not 400\\xa0mg, in ALS were described, and responsiveness of patients on motor function was observed. Since this was an open-label trial within a short period with a limited number of patients, further clinical trials will be required. Funding:\\nAMED and iPS Cell Research Fund.'}]}"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pubmed.build_pubmed_json('data/pmid-xmls/36467452.xml')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "folder = \"data/pmid-xmls\"\n",
    "file_list = os.listdir(folder)\n",
    "file_list = [\n",
    "    os.path.join(folder, file_name)\n",
    "    for file_name in file_list\n",
    "    if file_name.endswith(\".xml\")\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'data/pmid-xmls\\\\36464853.xml'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "file_list[36]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'abstract': [{'text': '\\n\\nBACKGROUND:\\nIn the phase 2 double-blind Study 211, a starting dose of lenvatinib 18\\u2009mg/day was compared with the approved starting dose of 24\\u2009mg/day in patients with radioiodine-refractory differentiated thyroid cancer (RR-DTC). Predefined criteria for noninferiority for efficacy in the 18\\u2009mg arm were not met; safety was similar in both arms. Impact of lenvatinib treatment on health-related quality-of-life (HRQoL) was a secondary endpoint of Study 211.\\n\\n\\nMETHODS:\\nPatients with RR-DTC were randomly assigned to a blinded starting dose of lenvatinib 18\\u2009mg/day or 24\\u2009mg/day. HRQoL was assessed at baseline, every 8\\u2009weeks until Week 24, then every 16\\u2009weeks, and at the off-treatment visit, using the EQ-5D-3L and FACT-G instruments. Completion and compliance rates, mean change from baseline, and times to first and definitive deterioration were evaluated.\\n\\n\\nRESULTS:\\nBaseline EQ-5D and FACT-G scores, and overall changes from baseline, were comparable between patients in the lenvatinib 18\\u2009mg/day (n\\xa0=\\u200977) and 24\\u2009mg/day arms (n\\xa0=\\xa075). For the 18\\u2009mg versus 24\\u2009mg arms, least squares mean differences were\\u2009-0.42 (95% CI -4.88, 4.03) for EQ-5D-VAS and 0.47 (95% CI -3.45, 4.39) for FACT-G total. Time to first deterioration did not significantly favor either arm; EQ-5D-VAS HR [18\\u2009mg/24\\u2009mg] 0.93 (95% CI 0.61-1.40), EQ-5D-HUI HR [18\\u2009mg/24\\u2009mg] 0.68 (95% CI 0.44-1.05), FACT-G total HR [18\\u2009mg/24\\u2009mg] 0.73 (95% CI 0.48-1.12). Time to definitive deterioration did not significantly favor either arm, though EQ-5D-VAS showed a trend in favor of the 24\\u2009mg arm (HR [18\\u2009mg/24\\u2009mg] 1.72; 95% CI 0.99-3.01); EQ-5D-HUI HR [18\\u2009mg/24\\u2009mg] was 0.96 (95% CI 0.57-1.63), FACT-G total HR [18\\u2009mg/24\\u2009mg] was 0.72 (95% CI 0.43-1.21).\\n\\n\\nCONCLUSIONS:\\nIn Study 211, HRQoL for patients in the lenvatinib 18\\u2009mg/day arm was not statistically different from that of patients in the 24\\u2009mg/day arm. These data further support the use of the approved lenvatinib starting dose of 24\\u2009mg/day in patients with RR-DTC.\\n\\n\\nCLINICALTRIALS:\\n\\n\\n\\nGOV NUMBER:\\nNCT02702388.',\n",
       "   'section_title': 'Abstract',\n",
       "   'section_type': 'ABSTRACT'}],\n",
       " 'meta_info': {'articleids': [{'idtype': 'pubmed', 'value': '36464853'},\n",
       "   {'idtype': 'doi', 'value': '10.1002/cam4.5308'}],\n",
       "  'kwd': 'HRQoL, dose intensity, kinase inhibitor, lenvatinib, quality of life, radioiodine-refractory differentiated thyroid cancer',\n",
       "  'dates_history': [{'date_type': 'revised',\n",
       "    'year': '2022',\n",
       "    'month': '09',\n",
       "    'day': '14'},\n",
       "   {'date_type': 'received', 'year': '2022', 'month': '05', 'day': '24'},\n",
       "   {'date_type': 'accepted', 'year': '2022', 'month': '09', 'day': '20'},\n",
       "   {'date_type': 'entrez', 'year': '2022', 'month': '12', 'day': '5'},\n",
       "   {'date_type': 'pubmed', 'year': '2022', 'month': '12', 'day': '6'},\n",
       "   {'date_type': 'medline', 'year': '2022', 'month': '12', 'day': '6'}],\n",
       "  'fulljournalname': 'Cancer medicine',\n",
       "  'journal_abbrev': 'Cancer Med',\n",
       "  'issn': '2045-7634',\n",
       "  'pubdate': '2022 Dec 04',\n",
       "  'title': 'Health-related quality-of-life analyses from a multicenter, randomized, double-blind phase 2 study of patients with differentiated thyroid cancer treated with lenvatinib 18 or 24\\u2009mg/day.',\n",
       "  'languages': 'eng',\n",
       "  'volume': '',\n",
       "  'issue': '',\n",
       "  'pages': '',\n",
       "  'authors': [{'first': 'Matthew H',\n",
       "    'middle': '',\n",
       "    'last': 'Taylor',\n",
       "    'suffix': '',\n",
       "    'initials': 'MH',\n",
       "    'affiliation': 'Earle A. Chiles Research Institute, Providence Portland Medical Center, Portland, Oregon, USA.',\n",
       "    'email': ''},\n",
       "   {'first': 'Sophie',\n",
       "    'middle': '',\n",
       "    'last': 'Leboulleux',\n",
       "    'suffix': '',\n",
       "    'initials': 'S',\n",
       "    'affiliation': 'Department of Nuclear Medicine and Endocrine Oncology, Gustave Roussy and University Paris Saclay, Villejuif, France.',\n",
       "    'email': ''},\n",
       "   {'first': 'Yury',\n",
       "    'middle': '',\n",
       "    'last': 'Panaseykin',\n",
       "    'suffix': '',\n",
       "    'initials': 'Y',\n",
       "    'affiliation': 'A. Tsyb Medical Radiological Research Center, branch of the NMRС of Radiology, Obninsk, Russian Federation.',\n",
       "    'email': ''},\n",
       "   {'first': 'Bhavana',\n",
       "    'middle': '',\n",
       "    'last': 'Konda',\n",
       "    'suffix': '',\n",
       "    'initials': 'B',\n",
       "    'affiliation': 'Division of Medical Oncology, The Ohio State University Comprehensive Cancer Center, Ohio, Columbus, USA.',\n",
       "    'email': ''},\n",
       "   {'first': 'Christelle',\n",
       "    'middle': '',\n",
       "    'last': 'de La Fouchardiere',\n",
       "    'suffix': '',\n",
       "    'initials': 'C',\n",
       "    'affiliation': 'Medical Oncology Department, Centre Léon Bérard, Lyon, France.',\n",
       "    'email': ''},\n",
       "   {'first': 'Brett G M',\n",
       "    'middle': '',\n",
       "    'last': 'Hughes',\n",
       "    'suffix': '',\n",
       "    'initials': 'BGM',\n",
       "    'affiliation': \"Department of Cancer Care Services, Royal Brisbane and Women's Hospital, University of Queensland, Queensland, Australia.\",\n",
       "    'email': ''},\n",
       "   {'first': 'Andrew G',\n",
       "    'middle': '',\n",
       "    'last': 'Gianoukakis',\n",
       "    'suffix': '',\n",
       "    'initials': 'AG',\n",
       "    'affiliation': 'The Lundquist Institute at Harbor-UCLA Medical Center, David Geffen School of Medicine at UCLA, California, Los Angeles/Torrance, USA.',\n",
       "    'email': ''},\n",
       "   {'first': 'Young Joo',\n",
       "    'middle': '',\n",
       "    'last': 'Park',\n",
       "    'suffix': '',\n",
       "    'initials': 'YJ',\n",
       "    'affiliation': 'Department of Internal Medicine, Seoul National University College of Medicine, Seoul, Republic of Korea.',\n",
       "    'email': ''},\n",
       "   {'first': 'Ilia',\n",
       "    'middle': '',\n",
       "    'last': 'Romanov',\n",
       "    'suffix': '',\n",
       "    'initials': 'I',\n",
       "    'affiliation': 'Department of Head & Neck Tumors, N.N. Blokhin Russian Cancer Research Center, Moscow, Russian Federation.',\n",
       "    'email': ''},\n",
       "   {'first': 'Monika K',\n",
       "    'middle': '',\n",
       "    'last': 'Krzyzanowska',\n",
       "    'suffix': '',\n",
       "    'initials': 'MK',\n",
       "    'affiliation': 'Department of Medical Oncology & Hematology, Princess Margaret Cancer Centre, Ontario, Toronto, Canada.',\n",
       "    'email': ''},\n",
       "   {'first': 'Diana',\n",
       "    'middle': '',\n",
       "    'last': 'Garbinsky',\n",
       "    'suffix': '',\n",
       "    'initials': 'D',\n",
       "    'affiliation': 'RTI Health Solutions, Research Triangle Park, North Carolina, USA.',\n",
       "    'email': ''},\n",
       "   {'first': 'Bintu',\n",
       "    'middle': '',\n",
       "    'last': 'Sherif',\n",
       "    'suffix': '',\n",
       "    'initials': 'B',\n",
       "    'affiliation': 'RTI Health Solutions, Research Triangle Park, North Carolina, USA.',\n",
       "    'email': ''},\n",
       "   {'first': 'Jie Janice',\n",
       "    'middle': '',\n",
       "    'last': 'Pan',\n",
       "    'suffix': '',\n",
       "    'initials': 'JJ',\n",
       "    'affiliation': 'Global Value and Access (GV&A), Oncology, Eisai Inc., New Jersey, Nutley, USA.',\n",
       "    'email': ''},\n",
       "   {'first': 'Terri A',\n",
       "    'middle': '',\n",
       "    'last': 'Binder',\n",
       "    'suffix': '',\n",
       "    'initials': 'TA',\n",
       "    'affiliation': 'Oncology Clinical Research, Eisai Inc., New Jersey, Nutley, USA.',\n",
       "    'email': ''},\n",
       "   {'first': 'Nicholas',\n",
       "    'middle': '',\n",
       "    'last': 'Sauter',\n",
       "    'suffix': '',\n",
       "    'initials': 'N',\n",
       "    'affiliation': 'Oncology Clinical Research, Eisai Inc., New Jersey, Nutley, USA.',\n",
       "    'email': ''},\n",
       "   {'first': 'Ran',\n",
       "    'middle': '',\n",
       "    'last': 'Xie',\n",
       "    'suffix': '',\n",
       "    'initials': 'R',\n",
       "    'affiliation': 'Biostatistics, Eisai Inc., New Jersey, Nutley, USA.',\n",
       "    'email': ''},\n",
       "   {'first': 'Marcia S',\n",
       "    'middle': '',\n",
       "    'last': 'Brose',\n",
       "    'suffix': '',\n",
       "    'initials': 'MS',\n",
       "    'affiliation': 'Department of Medical Oncology, Sidney Kimmel Cancer Center, Jefferson University (previous affiliation: Department of Otorhinolaryngology: Head and Neck Surgery, Abramson Cancer Center, University of Pennsylvania), Pennsylvania, Philadelphia, USA.',\n",
       "    'email': ''}],\n",
       "  'mesh_terms': '',\n",
       "  'citation_count': 2},\n",
       " 'journal_ranking': {'Journal': 'Cancer Medicine',\n",
       "  'Articles': '4.4K',\n",
       "  'Citations': '60.1K',\n",
       "  'Impact Factor': '4.8',\n",
       "  'h-index': '78'}}"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pubmed.build_pubmed_json(file_list[36])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# articles with more than one splits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "folder = \"data/pubmed\"\n",
    "file_list = os.listdir(folder)\n",
    "file_list = [\n",
    "    os.path.join(folder, file_name)\n",
    "    for file_name in file_list\n",
    "    if file_name.endswith(\".json\")\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 9461/9461 [00:00<00:00, 15286.99it/s]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "754"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from tqdm import tqdm\n",
    "\n",
    "\n",
    "article_list = []\n",
    "for file in tqdm(file_list):\n",
    "    # print(file)\n",
    "    with open(file, \"r\") as f:\n",
    "        data = json.load(f)\n",
    "    if len(data[\"article_splits\"]) > 1:\n",
    "        article_list.append(data)\n",
    "\n",
    "len(article_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'abstract': [{'text': '\\n\\nBACKGROUND:\\nIn the phase 2 double-blind Study 211, a starting dose of lenvatinib 18\\u2009mg/day was compared with the approved starting dose of 24\\u2009mg/day in patients with radioiodine-refractory differentiated thyroid cancer (RR-DTC). Predefined criteria for noninferiority for efficacy in the 18\\u2009mg arm were not met; safety was similar in both arms. Impact of lenvatinib treatment on health-related quality-of-life (HRQoL) was a secondary endpoint of Study 211.\\n\\n\\nMETHODS:\\nPatients with RR-DTC were randomly assigned to a blinded starting dose of lenvatinib 18\\u2009mg/day or 24\\u2009mg/day. HRQoL was assessed at baseline, every 8\\u2009weeks until Week 24, then every 16\\u2009weeks, and at the off-treatment visit, using the EQ-5D-3L and FACT-G instruments. Completion and compliance rates, mean change from baseline, and times to first and definitive deterioration were evaluated.\\n\\n\\nRESULTS:\\nBaseline EQ-5D and FACT-G scores, and overall changes from baseline, were comparable between patients in the lenvatinib 18\\u2009mg/day (n\\xa0=\\u200977) and 24\\u2009mg/day arms (n\\xa0=\\xa075). For the 18\\u2009mg versus 24\\u2009mg arms, least squares mean differences were\\u2009-0.42 (95% CI -4.88, 4.03) for EQ-5D-VAS and 0.47 (95% CI -3.45, 4.39) for FACT-G total. Time to first deterioration did not significantly favor either arm; EQ-5D-VAS HR [18\\u2009mg/24\\u2009mg] 0.93 (95% CI 0.61-1.40), EQ-5D-HUI HR [18\\u2009mg/24\\u2009mg] 0.68 (95% CI 0.44-1.05), FACT-G total HR [18\\u2009mg/24\\u2009mg] 0.73 (95% CI 0.48-1.12). Time to definitive deterioration did not significantly favor either arm, though EQ-5D-VAS showed a trend in favor of the 24\\u2009mg arm (HR [18\\u2009mg/24\\u2009mg] 1.72; 95% CI 0.99-3.01); EQ-5D-HUI HR [18\\u2009mg/24\\u2009mg] was 0.96 (95% CI 0.57-1.63), FACT-G total HR [18\\u2009mg/24\\u2009mg] was 0.72 (95% CI 0.43-1.21).\\n\\n\\nCONCLUSIONS:\\nIn Study 211, HRQoL for patients in the lenvatinib 18\\u2009mg/day arm was not statistically different from that of patients in the 24\\u2009mg/day arm. These data further support the use of the approved lenvatinib starting dose of 24\\u2009mg/day in patients with RR-DTC.\\n\\n\\nCLINICALTRIALS:\\n\\n\\n\\nGOV NUMBER:\\nNCT02702388.',\n",
       "   'section_title': 'Abstract',\n",
       "   'section_type': 'ABSTRACT'}],\n",
       " 'meta_info': {'articleids': [{'idtype': 'pubmed', 'value': '36464853'},\n",
       "   {'idtype': 'doi', 'value': '10.1002/cam4.5308'}],\n",
       "  'kwd': 'HRQoL, dose intensity, kinase inhibitor, lenvatinib, quality of life, radioiodine-refractory differentiated thyroid cancer',\n",
       "  'dates_history': [{'date_type': 'revised',\n",
       "    'year': '2022',\n",
       "    'month': '09',\n",
       "    'day': '14'},\n",
       "   {'date_type': 'received', 'year': '2022', 'month': '05', 'day': '24'},\n",
       "   {'date_type': 'accepted', 'year': '2022', 'month': '09', 'day': '20'},\n",
       "   {'date_type': 'entrez', 'year': '2022', 'month': '12', 'day': '5'},\n",
       "   {'date_type': 'pubmed', 'year': '2022', 'month': '12', 'day': '6'},\n",
       "   {'date_type': 'medline', 'year': '2022', 'month': '12', 'day': '6'}],\n",
       "  'fulljournalname': 'Cancer medicine',\n",
       "  'journal_abbrev': 'Cancer Med',\n",
       "  'issn': '2045-7634',\n",
       "  'pubdate': '2022 Dec 04',\n",
       "  'title': 'Health-related quality-of-life analyses from a multicenter, randomized, double-blind phase 2 study of patients with differentiated thyroid cancer treated with lenvatinib 18 or 24\\u2009mg/day.',\n",
       "  'languages': 'eng',\n",
       "  'volume': '',\n",
       "  'issue': '',\n",
       "  'pages': '',\n",
       "  'authors': [{'first': 'Matthew H',\n",
       "    'middle': '',\n",
       "    'last': 'Taylor',\n",
       "    'suffix': '',\n",
       "    'initials': 'MH',\n",
       "    'affiliation': 'Earle A. Chiles Research Institute, Providence Portland Medical Center, Portland, Oregon, USA.',\n",
       "    'email': ''},\n",
       "   {'first': 'Sophie',\n",
       "    'middle': '',\n",
       "    'last': 'Leboulleux',\n",
       "    'suffix': '',\n",
       "    'initials': 'S',\n",
       "    'affiliation': 'Department of Nuclear Medicine and Endocrine Oncology, Gustave Roussy and University Paris Saclay, Villejuif, France.',\n",
       "    'email': ''},\n",
       "   {'first': 'Yury',\n",
       "    'middle': '',\n",
       "    'last': 'Panaseykin',\n",
       "    'suffix': '',\n",
       "    'initials': 'Y',\n",
       "    'affiliation': 'A. Tsyb Medical Radiological Research Center, branch of the NMRС of Radiology, Obninsk, Russian Federation.',\n",
       "    'email': ''},\n",
       "   {'first': 'Bhavana',\n",
       "    'middle': '',\n",
       "    'last': 'Konda',\n",
       "    'suffix': '',\n",
       "    'initials': 'B',\n",
       "    'affiliation': 'Division of Medical Oncology, The Ohio State University Comprehensive Cancer Center, Ohio, Columbus, USA.',\n",
       "    'email': ''},\n",
       "   {'first': 'Christelle',\n",
       "    'middle': '',\n",
       "    'last': 'de La Fouchardiere',\n",
       "    'suffix': '',\n",
       "    'initials': 'C',\n",
       "    'affiliation': 'Medical Oncology Department, Centre Léon Bérard, Lyon, France.',\n",
       "    'email': ''},\n",
       "   {'first': 'Brett G M',\n",
       "    'middle': '',\n",
       "    'last': 'Hughes',\n",
       "    'suffix': '',\n",
       "    'initials': 'BGM',\n",
       "    'affiliation': \"Department of Cancer Care Services, Royal Brisbane and Women's Hospital, University of Queensland, Queensland, Australia.\",\n",
       "    'email': ''},\n",
       "   {'first': 'Andrew G',\n",
       "    'middle': '',\n",
       "    'last': 'Gianoukakis',\n",
       "    'suffix': '',\n",
       "    'initials': 'AG',\n",
       "    'affiliation': 'The Lundquist Institute at Harbor-UCLA Medical Center, David Geffen School of Medicine at UCLA, California, Los Angeles/Torrance, USA.',\n",
       "    'email': ''},\n",
       "   {'first': 'Young Joo',\n",
       "    'middle': '',\n",
       "    'last': 'Park',\n",
       "    'suffix': '',\n",
       "    'initials': 'YJ',\n",
       "    'affiliation': 'Department of Internal Medicine, Seoul National University College of Medicine, Seoul, Republic of Korea.',\n",
       "    'email': ''},\n",
       "   {'first': 'Ilia',\n",
       "    'middle': '',\n",
       "    'last': 'Romanov',\n",
       "    'suffix': '',\n",
       "    'initials': 'I',\n",
       "    'affiliation': 'Department of Head & Neck Tumors, N.N. Blokhin Russian Cancer Research Center, Moscow, Russian Federation.',\n",
       "    'email': ''},\n",
       "   {'first': 'Monika K',\n",
       "    'middle': '',\n",
       "    'last': 'Krzyzanowska',\n",
       "    'suffix': '',\n",
       "    'initials': 'MK',\n",
       "    'affiliation': 'Department of Medical Oncology & Hematology, Princess Margaret Cancer Centre, Ontario, Toronto, Canada.',\n",
       "    'email': ''},\n",
       "   {'first': 'Diana',\n",
       "    'middle': '',\n",
       "    'last': 'Garbinsky',\n",
       "    'suffix': '',\n",
       "    'initials': 'D',\n",
       "    'affiliation': 'RTI Health Solutions, Research Triangle Park, North Carolina, USA.',\n",
       "    'email': ''},\n",
       "   {'first': 'Bintu',\n",
       "    'middle': '',\n",
       "    'last': 'Sherif',\n",
       "    'suffix': '',\n",
       "    'initials': 'B',\n",
       "    'affiliation': 'RTI Health Solutions, Research Triangle Park, North Carolina, USA.',\n",
       "    'email': ''},\n",
       "   {'first': 'Jie Janice',\n",
       "    'middle': '',\n",
       "    'last': 'Pan',\n",
       "    'suffix': '',\n",
       "    'initials': 'JJ',\n",
       "    'affiliation': 'Global Value and Access (GV&A), Oncology, Eisai Inc., New Jersey, Nutley, USA.',\n",
       "    'email': ''},\n",
       "   {'first': 'Terri A',\n",
       "    'middle': '',\n",
       "    'last': 'Binder',\n",
       "    'suffix': '',\n",
       "    'initials': 'TA',\n",
       "    'affiliation': 'Oncology Clinical Research, Eisai Inc., New Jersey, Nutley, USA.',\n",
       "    'email': ''},\n",
       "   {'first': 'Nicholas',\n",
       "    'middle': '',\n",
       "    'last': 'Sauter',\n",
       "    'suffix': '',\n",
       "    'initials': 'N',\n",
       "    'affiliation': 'Oncology Clinical Research, Eisai Inc., New Jersey, Nutley, USA.',\n",
       "    'email': ''},\n",
       "   {'first': 'Ran',\n",
       "    'middle': '',\n",
       "    'last': 'Xie',\n",
       "    'suffix': '',\n",
       "    'initials': 'R',\n",
       "    'affiliation': 'Biostatistics, Eisai Inc., New Jersey, Nutley, USA.',\n",
       "    'email': ''},\n",
       "   {'first': 'Marcia S',\n",
       "    'middle': '',\n",
       "    'last': 'Brose',\n",
       "    'suffix': '',\n",
       "    'initials': 'MS',\n",
       "    'affiliation': 'Department of Medical Oncology, Sidney Kimmel Cancer Center, Jefferson University (previous affiliation: Department of Otorhinolaryngology: Head and Neck Surgery, Abramson Cancer Center, University of Pennsylvania), Pennsylvania, Philadelphia, USA.',\n",
       "    'email': ''}],\n",
       "  'mesh_terms': ''},\n",
       " 'pmid': '36464853',\n",
       " 'article_splits': [{'PMID36464853_abstract_split_0': 'BACKGROUND:\\nIn the phase 2 double-blind Study 211, a starting dose of lenvatinib 18\\u2009mg/day was compared with the approved starting dose of 24\\u2009mg/day in patients with radioiodine-refractory differentiated thyroid cancer (RR-DTC). Predefined criteria for noninferiority for efficacy in the 18\\u2009mg arm were not met; safety was similar in both arms. Impact of lenvatinib treatment on health-related quality-of-life (HRQoL) was a secondary endpoint of Study 211. METHODS:\\nPatients with RR-DTC were randomly assigned to a blinded starting dose of lenvatinib 18\\u2009mg/day or 24\\u2009mg/day. HRQoL was assessed at baseline, every 8\\u2009weeks until Week 24, then every 16\\u2009weeks, and at the off-treatment visit, using the EQ-5D-3L and FACT-G instruments. Completion and compliance rates, mean change from baseline, and times to first and definitive deterioration were evaluated. RESULTS:\\nBaseline EQ-5D and FACT-G scores, and overall changes from baseline, were comparable between patients in the lenvatinib 18\\u2009mg/day (n\\xa0=\\u200977) and 24\\u2009mg/day arms (n\\xa0=\\xa075). For the 18\\u2009mg versus 24\\u2009mg arms, least squares mean differences were\\u2009-0.42 (95% CI -4.88, 4.03) for EQ-5D-VAS and 0.47 (95% CI -3.45, 4.39) for FACT-G total. Time to first deterioration did not significantly favor either arm; EQ-5D-VAS HR [18\\u2009mg/24\\u2009mg] 0.93 (95% CI 0.61-1.40), EQ-5D-HUI HR [18\\u2009mg/24\\u2009mg] 0.68 (95% CI 0.44-1.05), FACT-G total HR [18\\u2009mg/24\\u2009mg] 0.73 (95% CI 0.48-1.12).'},\n",
       "  {'PMID36464853_abstract_split_1': 'Time to first deterioration did not significantly favor either arm; EQ-5D-VAS HR [18\\u2009mg/24\\u2009mg] 0.93 (95% CI 0.61-1.40), EQ-5D-HUI HR [18\\u2009mg/24\\u2009mg] 0.68 (95% CI 0.44-1.05), FACT-G total HR [18\\u2009mg/24\\u2009mg] 0.73 (95% CI 0.48-1.12). Time to definitive deterioration did not significantly favor either arm, though EQ-5D-VAS showed a trend in favor of the 24\\u2009mg arm (HR [18\\u2009mg/24\\u2009mg] 1.72; 95% CI 0.99-3.01); EQ-5D-HUI HR [18\\u2009mg/24\\u2009mg] was 0.96 (95% CI 0.57-1.63), FACT-G total HR [18\\u2009mg/24\\u2009mg] was 0.72 (95% CI 0.43-1.21). CONCLUSIONS:\\nIn Study 211, HRQoL for patients in the lenvatinib 18\\u2009mg/day arm was not statistically different from that of patients in the 24\\u2009mg/day arm. These data further support the use of the approved lenvatinib starting dose of 24\\u2009mg/day in patients with RR-DTC. CLINICALTRIALS:\\n\\n\\n\\nGOV NUMBER:\\nNCT02702388.'}]}"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "article_list[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
